### PR TITLE
[WIP][Feature] support ConvViT, the backbone arch of MCMAE (previously ConvMAE)

### PR DESCRIPTION
**Pull request by linziyi96**
This pull request adds a new backbone architecture ConvViT, a mixed convolutional-transformer architecture used in MCMAE (previously ConvMAE, https://openreview.net/forum?id=qm5LpHyyOUO) that supports MAE pretraining.

Accuracy comparison (ImageNet-1k, ConvMAE-B fine-tuned. acc1, acc5 are using our implementation and the converted weights; acc1_ref, acc5_ref are using the implementation and weights from https://github.com/Alpha-VL/ConvMAE):
tot=50000, acc1=84.90%, acc5=97.11%, acc1_ref=84.98%, acc5_ref=97.14%

TODO: add test cases, example configuration and model card.

### CODE REVIEW

File: mmcls/models/backbones/convvit.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+                 embed_dims: int,
+                 feedforward_channels: int,
+                 drop_rate: float = 0.,
+                 act_cfg: dict = dict(type='GELU'),
+                 init_cfg: Optional[dict] = None):
+        super().__init__(init_cfg=init_cfg)
+
+        self.fc1 = nn.Conv2d(embed_dims, feedforward_channels, 1)
+        self.act = build_activation_layer(act_cfg)
+        self.dropout = (
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:41]:
```suggestion
        self.dropout = 
```

File: mmcls/models/backbones/convvit.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+            nn.Dropout(drop_rate) if drop_rate > 0. else nn.Identity())
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:42]:
```suggestion
            nn.Dropout(drop_rate) if drop_rate > 0. else nn.Identity()
```

File: mmcls/models/backbones/convvit.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+        embed_dims (int): the feature dimension
+        feedforward_channels (int): the MLP hidden feature dimension
+        drop_rate (float): probability of an element to be zeroed between
+            two fully-connected layers. defaults to 0.
+        act_cfg (dict): config for the activation between the two 
+            fully-connected layers. defaults to ``dict(type='GELU')``.
+        init_cfg (dict, optional): initialization config dict. defaults to None.
+    """
+
+    def __init__(self,
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:31]: There should be return typehint for the ``init`` func

File: mmcls/models/backbones/convvit.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+                 init_cfg: Optional[dict] = None):
+        super().__init__(init_cfg=init_cfg)
+
+        self.fc1 = nn.Conv2d(embed_dims, feedforward_channels, 1)
+        self.act = build_activation_layer(act_cfg)
+        self.dropout = (
+            nn.Dropout(drop_rate) if drop_rate > 0. else nn.Identity())
+        self.fc2 = nn.Conv2d(feedforward_channels, embed_dims, 1)
+
+    def forward(self, x):
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:45]: missing return typehint and type hint for ``x``

File: mmcls/models/backbones/convvit.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+from mmcv.runner.base_module import BaseModule
+
+from ..utils import to_2tuple
+from ..builder import BACKBONES
+from .base_backbone import BaseBackbone
+from .vision_transformer import TransformerEncoderLayer
+
+
+class ConvMlp(BaseModule):
+    """Implements the Mlp part for ConvBlock. The main difference is ConvMlp
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:18]: ``The main difference is ConvMlp implies N, C, H, W input format.`` is confused, since it does not specifiy the module ``ConvMlp`` is compared to.

File: mmcls/models/backbones/convvit.py [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    """
+
+    def __init__(self,
+                 embed_dims: int,
+                 feedforward_channels: int,
+                 drop_rate: float = 0.,
+                 drop_path_rate: float = 0.,
+                 act_cfg: dict = dict(type='GELU'),
+                 norm_cfg: dict = dict(type='LN'),
+                 init_cfg=None):
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:75]: typehint for ``init_cfg`` and the return typehint

File: mmcls/models/backbones/convvit.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+    @property
+    def norm1(self):
+        return getattr(self, self.norm1_name)
+
+    @property
+    def norm2(self):
+        return getattr(self, self.norm2_name)
+
+    def forward(self, x):
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:110]: typehint

File: mmcls/models/backbones/convvit.py [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+        x_id = x
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:111]: ``x_id`` -> ``x_identity``

File: mmcls/models/backbones/convvit.py [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+    def forward(self, x):
+        outs = []
+
+        for stage_idx in range(len(self.blocks)):
+            stage_type = 'c' if stage_idx < 2 else 't'
+
+            x, stage_out_size = self.patch_embeds[stage_idx](x)
+            x = self.patch_embed_act(x)
+            def convert_nlc_to_nchw(x):
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:351]: It is recommended rewriting this func as a private func of ``ConvViT``

File: mmcls/models/backbones/convvit.py [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+            print_log(
+                f'Resize pos_embed.{i} shape from {ckpt_pos_embed_shape} '
+                f'to {self.pos_embeds[i].shape}.',
+                logger=logger)
+            state_dict[name] = resize_pos_embed(state_dict[name],
+                ckpt_pos_embed_shape[1:-1],
+                self.pos_embeds[i].shape[1:-1],
+                self.interpolate_mode, 0)
+
+    def forward(self, x):
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:343]: typehint and return typehint, please also check other parts of this file

File: mmcls/models/backbones/convvit.py [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+        outs = []
+
+        for stage_idx in range(len(self.blocks)):
+            stage_type = 'c' if stage_idx < 2 else 't'
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:347]: Maybe ``c`` denotes ``convolution`` and ``t``denotes ``transformer``, I guesss. Better choose another more explicit name.

File: mmcls/models/backbones/convvit.py [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+                        torch.zeros(1, *stage_img_size, embed_dims[stage_idx]))
+                else: # conv stage
+                    layer_class = ConvBlock
+                    layer_conv_norm_cfg = conv_norm_cfg.copy()
+                    _layer_cfg.update(
+                        norm_cfg=layer_conv_norm_cfg,
+                        )
+                _layer_cfg.update(layer_cfgs[global_layer_idx])
+                blocks.append(layer_class(**_layer_cfg))
+            self.blocks.append(nn.ModuleList(blocks))
```
Comment (YuanLiuuuuuu) [inline mmcls/models/backbones/convvit.py:296]: Please use ``self.layers``, to be consistent with ``VisionTransformer`` in MMCls.

**Comment by linziyi96** [comment]
Evaluation accuracy verified (using tools/dist_test.sh: 84.9 reproduced vs. 85.0 claimed on in1k val, loading released checkpoint trained by the original implementation).

TODO: verify fine-tuning configuration.
